File: Tools/ADAccountManager/ADSettingsParser.cs
/// <summary>
///   Solution : NovaTend
///   Project : ADAccountManager.dll
///   Module : ADSettingsParser.cs
///   Description :  implements the methods of class ADSettingsParser
/// </summary>

using System;
using System.Text.RegularExpressions;
using System.Collections.Generic;
using System.Linq;
using System.Text;

namespace ADAccountManagement
{
    /// <summary>
    /// Parser for AD settings string
    /// 
    /// Example of AD settings string: <domain: DC1.LOCAL><defaultOU: OU = WorkUsers, DC=DC1, DC=local>
    /// <defaultRootOU: DC=dc1, DC=local><sUserName:peter><sUserPassword: 4pk2work>
    /// 
    /// Example of AD group settings string: <groupScope: global><groupType: security>
    /// 
    /// </summary>
    public class ADSettingsParser
    {

        /// <summary>
        /// Read one field of AD settings from the input settings string
        /// </summary>
        /// <param name="settingsString">settings string</param>
        /// <returns>AD settings field string</returns>
        private string GetField(string settingsString)
        {
            string fieldString = "";

            int beginPos = settingsString.IndexOf("<");
            int endPos = settingsString.IndexOf(">");

            if ((beginPos != -1) && (endPos != -1) && (beginPos < endPos))
            {
                int fieldLength = endPos - beginPos + 1;
                fieldString = settingsString.Substring(beginPos, fieldLength);
            }

            return fieldString;
        }

        /// <summary>
        /// Read one key marked field of AD settings 
        /// from the input settings string
        /// </summary>
        /// <param name="settingsString">settings string</param>
        /// <param name="keyString">key</param>
        /// <returns>AD settings field string</returns>
        private string GetFieldByKey(string settingsString, string keyString)
        {
            string _settingsString = settingsString;
            _settingsString = Regex.Replace(_settingsString, @"\s+", "");

            string valueString = "";
            bool found = false;
            int pos = -1;

            do
            {
                string fieldString = GetField(_settingsString);
                pos = CheckField(fieldString, keyString);

                if (pos != -1)
                {
                    pos += keyString.Length;
                    valueString = fieldString.Substring(pos);
                    valueString = valueString.Remove(valueString.Length - 1, 1);

                    found = true;
                }
                else
                {
                    pos = _settingsString.IndexOf(fieldString);

                    if (pos != -1)
                        _settingsString = _settingsString.Remove(pos, fieldString.Length);
                }
                                
            } while ((!found) && (_settingsString.Length > 0));

            return valueString;
        }

        /// <summary>
        /// Check field settings string. Compare it with
        /// key marked field string
        /// </summary>
        /// <param name="fieldString">field settings string</param>
        /// <param name="fieldKey">key</param>
        /// <returns>index of key</returns>
        private int CheckField(string fieldString, string fieldKey)
        {
            return fieldString.IndexOf(fieldKey);
        }

        /// <summary>
        /// Get AD service user name
        /// </summary>
        /// <param name="settingsString">settings string</param>
        /// <returns>service user name</returns>
        public string GetServiceUserName(string settingsString)
        {
            return GetFieldByKey(settingsString, "sUserName:");
        }

        /// <summary>
        /// Get AD service user password
        /// </summary>
        /// <param name="settingsString">settings string</param>
        /// <returns>service user password</returns>
        public string GetServiceUserPassword(string settingsString)
        {
            return GetFieldByKey(settingsString, "sUserPassword:");
        }

        /// <summary>
        /// Get AD domain name
        /// </summary>
        /// <param name="settingsString">settings string</param>
        /// <returns>domain name</returns>
        public string GetDomainName(string settingsString)
        {
            return GetFieldByKey(settingsString, "domain:");
        }

        /// <summary>
        /// Get AD default OU for user accounts management
        /// </summary>
        /// <param name="settingsString">settings string</param>
        /// <returns>default OU</returns>
        public string GetDefaultOU(string settingsString)
        {
            return GetFieldByKey(settingsString, "defaultOU:");
        }

        /// <summary>
        /// Get AD default root OU for user accounts management
        /// </summary>
        /// <param name="settingsString">settings string</param>
        /// <returns>default root OU</returns>
        public string GetDefaultRootOU(string settingsString)
        {
            return GetFieldByKey(settingsString, "defaultRootOU:");
        }

        /// <summary>
        /// Get AD group scope
        /// </summary>
        /// <param name="groupSettingsString">group settings string</param>
        /// <returns>group scope name</returns>
        public string GetGroupScope(string groupSettingsString)
        {
            return GetFieldByKey(groupSettingsString, "groupScope:");
        }

        /// <summary>
        /// Get AD group type name
        /// </summary>
        /// <param name="groupSettingsString">group settings string</param>
        /// <returns>group type name string</returns>
        public string GetGroupTypeName(string groupSettingsString)
        {
            return GetFieldByKey(groupSettingsString, "groupType:");
        }
    }
}

File: Kernel/z-adman-01/ADSettingsParser.cpp
/****************************************************************

   Solution : NovaTend

   Project : z-adman-01.dll

   Module : ADSettingsParser.cpp

   Description : this module implements methods of
				  class ADSettingsParser

*****************************************************************/

#include "stdafx.h"
#include "ADSettingsParser.h"
#include <ContentDataDef.h>
#include <ctype.h>
#include <string.h>


// z-adman-01.dll definition structure
extern DLL_COMMON_DEFINITION zADManDllDefinition;

/****************************************************************

   Class : ADAccountManager

   Method : Constructor

   Parameters :
			Input : 
				dllUserID - ID of DLL user
				logWriter - LOG writer instance
				
*****************************************************************/
ADSettingsParser :: ADSettingsParser(unsigned long dllUserID, LogWriter * logWriter) : SystemBase(dllUserID, &zADManDllDefinition)
{
	this -> logWriter = logWriter;

	#if defined(_DEBUG_LOG_ALL) || defined(_DEBUG_LOG_ADMAN)
		WriteToLog("ADSettingsParser : constructor");
	#endif	
}

/****************************************************************

   Class : ADSettingsParser

   Method : Destructor

*****************************************************************/
ADSettingsParser :: ~ADSettingsParser(void)
{
	#if defined(_DEBUG_LOG_ALL) || defined(_DEBUG_LOG_ADMAN)
		WriteToLog("ADSettingsParser : destructor");
	#endif
}

/****************************************************************

   Class : ADSettingsParser

   Method : SubStringIndex

   Parameters :
			Input : 
				source - source string
				substr - sub string

	Return : substring index

	Description : Get index of substring

*****************************************************************/
int ADSettingsParser :: SubStringIndex(char * source, char * substr)
{
	int index = -1;
		
	if (source)
	{
		if (substr) 
		{
			char * str = strstr(source, substr);
			if ((str - source) >= 0)	
				index = str - source;
		}
	}

	return index;
}

/****************************************************************

   Class : ADSettingsParser

   Method : RemoveSubString

   Parameters :
			Input : 
				source - source string
				substr - sub string

	Return : result string

	Description : Remove substring

*****************************************************************/
char * ADSettingsParser :: RemoveSubString(char * source, char * substr)
{
	char dest[STR_VALUE_LEN];

	char * str = strstr(source, substr);

	if (str)
	{
		memcpy(dest, source, str - source);
		dest[str - source] = '\0';
		strcat(dest, str + strlen(substr));
		strcpy(source, dest);
		return source;
	}
	else
		return NULL;
}


/****************************************************************

   Class : ADSettingsParser

   Method : RemoveWhiteSpaces

   Parameters :
			Input : 
				source - source string
			Output:
				target - result string

	Return : result string

	Description : Remove all white spaces

*****************************************************************/
char * ADSettingsParser :: RemoveWhiteSpaces(char * source, char * target)
{
	#if defined(_DEBUG_LOG_ALL) || defined(_DEBUG_LOG_ADMAN)
		WriteToLog("RemoveWhiteSpaces : Start");
	#endif	

	if (!(source) || !(target)) 
		return NULL;
	
	int c = 0;

	for (unsigned int i = 0; i < strlen(source); i++)
	{
		if (!isspace(source[i]))
			target[c++] = source[i];
    }

	#if defined(_DEBUG_LOG_ALL) || defined(_DEBUG_LOG_ADMAN)
		WriteToLog("RemoveWhiteSpaces : Done");
	#endif	
	
	return target;
}

/****************************************************************

   Class : ADSettingsParser

   Method : GetField

   Parameters :
			Input : 
				settingsString - source settings string
			Output:
				fieldString - result field string

	Return : result string

	Description : Get field string

*****************************************************************/
char * ADSettingsParser :: GetField(char * settingsString, char * fieldString)
{
	#if defined(_DEBUG_LOG_ALL) || defined(_DEBUG_LOG_ADMAN)
		WriteToLog("GetField : Start");
	#endif	

	int beginPos = SubStringIndex(settingsString, "<");
	int endPos = SubStringIndex(settingsString, ">");

	if ((beginPos != -1) && (endPos != -1) && (beginPos < endPos))
    {
		int fieldLength = endPos - beginPos + 1;
		strncpy(fieldString, settingsString + beginPos, fieldLength);
		fieldString[fieldLength] = '\0';
	}

	#if defined(_DEBUG_LOG_ALL) || defined(_DEBUG_LOG_ADMAN)
		WriteToLog("GetField : Done");
	#endif	

	return fieldString;
}

/****************************************************************

   Class : ADSettingsParser

   Method : GetFiledByKey

   Parameters :
			Input : 
				settingsString - source settings string
				keyString - key string
			Output:
				fieldString - result field string

	Return : result string

	Description : Get key field string

*****************************************************************/
char * ADSettingsParser :: GetFieldByKey(char * settingsString, char * keyString, char * fieldString)
{
	#if defined(_DEBUG_LOG_ALL) || defined(_DEBUG_LOG_ADMAN)
		WriteToLog("GetFiledByKey : Start");
	#endif	

	char _settingsString[SETTINGS_STRING_LEN] = "";
	char valueString[STR_VALUE_LEN] = "";
	RemoveWhiteSpaces(settingsString, _settingsString);
		
    bool found = false;
    int pos = -1;

	do
    {
		GetField(_settingsString, fieldString);
        pos = CheckField(fieldString, keyString);

        if (pos != -1)
        {
			pos += strlen(keyString);
			int len = strlen(fieldString) - strlen(keyString) - 2;
			strncpy(valueString, fieldString + pos, len);
			strcpy(fieldString, valueString);
           	found = true;
        }
        else
        {
			RemoveSubString(_settingsString, fieldString);
        }
                               
	} while ((!found) && (strlen(_settingsString) > 0));

	#if defined(_DEBUG_LOG_ALL) || defined(_DEBUG_LOG_ADMAN)
		WriteToLog("GetFiledByKey : Done");
	#endif	

	return fieldString;
}

/****************************************************************

   Class : ADSettingsParser

   Method : CheckField

   Parameters :
			Input : 
				fieldString - source settings string
				fieldKey - result field string
			
	Return : substring index

	Description : Check field settings string. Compare it with key marked field string

*****************************************************************/
int ADSettingsParser :: CheckField(char * fieldString, char * fieldKey)
{
	return SubStringIndex(fieldString, fieldKey);
}

/****************************************************************

   Class : ADSettingsParser

   Method : GetServiceUserName

   Parameters :
			Input : 
				settingsString - source settings string
			Output:
				sUserName - user name

	Return : result string

	Description : Get AD service user name

*****************************************************************/
char * ADSettingsParser :: GetServiceUserName(char * settingsString, char * sUserName)
{
	return GetFieldByKey(settingsString, "sUserName:", sUserName);
}

/****************************************************************

   Class : ADSettingsParser

   Method : GetServiceUserPassword

   Parameters :
			Input : 
				settingsString - source settings string
			Output:
				sUserPassword - user password

	Return : result string

	Description : Get AD service user password

*****************************************************************/
char * ADSettingsParser :: GetServiceUserPassword(char * settingsString, char * sUserPassword)
{
	return GetFieldByKey(settingsString, "sUserPassword:", sUserPassword);
}

/****************************************************************

   Class : ADSettingsParser

   Method : GetDomainName

   Parameters :
			Input : 
				settingsString - source settings string
			Output:
				domainName - domain name

	Return : result string

	Description : Get AD domain name

*****************************************************************/
char * ADSettingsParser :: GetDomainName(char * settingsString, char * domainName)
{
	return GetFieldByKey(settingsString, "domain:", domainName);
}

/****************************************************************

   Class : ADSettingsParser

   Method : GetDefaultOU

   Parameters :
			Input : 
				settingsString - source settings string
			Output:
				defaultOU - default OU

	Return : result string

	Description : Get AD default OU

*****************************************************************/
char * ADSettingsParser :: GetDefaultOU(char * settingsString, char * defaultOU)
{
	return GetFieldByKey(settingsString, "defaultOU:", defaultOU);
}

/****************************************************************

   Class : ADSettingsParser

   Method : GetDefaultRootOU

   Parameters :
			Input : 
				settingsString - source settings string
			Output:
				defaultRootOU - default root OU

	Return : result string

	Description : Get AD default root OU

*****************************************************************/
char * ADSettingsParser :: GetDefaultRootOU(char * settingsString, char * defaultRootOU)
{
	return GetFieldByKey(settingsString, "defaultRootOU:", defaultRootOU);
}

/****************************************************************

   Class : ADSettingsParser

   Method : GetGroupScopeName

   Parameters :
			Input : 
				settingsString - source settings string
			Output:
				groupScope - group scope name

	Return : result string

	Description : Get AD group scope name

*****************************************************************/
char * ADSettingsParser :: GetGroupScopeName(char * settingsString, char * groupScope)
{
	return GetFieldByKey(settingsString, "groupScope:", groupScope);
}

/****************************************************************

   Class : ADSettingsParser

   Method : GetGroupTypeName

   Parameters :
			Input : 
				settingsString - source settings string
			Output:
				groupType - group type name

	Return : result string

	Description : Get AD group type name

*****************************************************************/
char * ADSettingsParser :: GetGroupTypeName(char * settingsString, char * groupType)
{
	return GetFieldByKey(settingsString, "groupType:", groupType);
}

/****************************************************************

   Class : ADSettingsParser

   Method : WriteToLog

   Parameters :
			Input : 
				message - message for writting into log-file

	Return : error code

	Description : Write message to log-file

*****************************************************************/
int  ADSettingsParser :: WriteToLog(char * message)
{
	if (logWriter)
		return logWriter -> WriteToLog(message);
	else
		return LOG_FILE_OPEN_ERROR;

}

